File: research/arekappa/integration/kappa_ir.py
"""Pure, non-production AREKappa IR helpers.

This module has no network, database, clock, random, or filesystem effects.
It accepts exact scaled integers only and emits canonical JSON bytes.
"""

from __future__ import annotations

from hashlib import sha256
import json
from typing import Any, Mapping, Sequence
import unicodedata

KAPPA_SCALE = 1_000_000


def normalize_weights_largest_remainder(weights: Sequence[int], scale: int = KAPPA_SCALE) -> tuple[int, ...]:
    """Normalize non-negative integer weights to exactly ``scale`` units.

    Ties are resolved by ascending index. No floating-point operation is used.
    """
    if not isinstance(scale, int) or isinstance(scale, bool) or scale <= 0:
        raise ValueError("scale must be a positive integer")
    values = tuple(weights)
    if not values:
        raise ValueError("weights must not be empty")
    if any(not isinstance(value, int) or isinstance(value, bool) or value < 0 for value in values):
        raise ValueError("weights must contain non-negative integers only")
    total = sum(values)
    if total <= 0:
        raise ValueError("weight sum must be positive")

    numerators = tuple(value * scale for value in values)
    floors = [numerator // total for numerator in numerators]
    remaining = scale - sum(floors)
    order = sorted(range(len(values)), key=lambda index: (-(numerators[index] % total), index))
    for index in order[:remaining]:
        floors[index] += 1
    return tuple(floors)


def matrix_vector_product(matrix: Sequence[Sequence[int]], vector: Sequence[int]) -> tuple[int, ...]:
    rows = tuple(tuple(row) for row in matrix)
    values = tuple(vector)
    if not rows or not values or any(len(row) != len(values) for row in rows):
        raise ValueError("matrix and vector dimensions must be non-empty and compatible")
    if any(not isinstance(value, int) or isinstance(value, bool) for row in rows for value in row):
        raise ValueError("matrix values must be integers")
    if any(not isinstance(value, int) or isinstance(value, bool) for value in values):
        raise ValueError("vector values must be integers")
    return tuple(sum(coefficient * value for coefficient, value in zip(row, values)) for row in rows)


def _normalize_json(value: Any) -> Any:
    if value is None or isinstance(value, (bool, int)):
        return value
    if isinstance(value, float):
        raise ValueError("floats are forbidden in canonical AREKappa JSON")
    if isinstance(value, str):
        return unicodedata.normalize("NFC", value)
    if isinstance(value, Mapping):
        normalized: dict[str, Any] = {}
        for key, child in value.items():
            if not isinstance(key, str):
                raise ValueError("canonical JSON object keys must be strings")
            normalized[unicodedata.normalize("NFC", key)] = _normalize_json(child)
        return normalized
    if isinstance(value, (list, tuple)):
        return [_normalize_json(child) for child in value]
    raise ValueError(f"unsupported canonical JSON value: {type(value).__name__}")


def canonical_json_bytes(value: Any) -> bytes:
    normalized = _normalize_json(value)
    return json.dumps(
        normalized,
        ensure_ascii=False,
        allow_nan=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode("utf-8")


def canonical_sha256(value: Any) -> str:
    return sha256(canonical_json_bytes(value)).hexdigest()

File: research/arekappa/integration/kappa-ir.ts
export const KAPPA_SCALE = 1_000_000n;

export function normalizeWeightsLargestRemainder(
  weights: readonly bigint[],
  scale: bigint = KAPPA_SCALE,
): readonly bigint[] {
  if (scale <= 0n) throw new Error("scale must be positive");
  if (weights.length === 0) throw new Error("weights must not be empty");
  if (weights.some((value) => value < 0n)) throw new Error("weights must be non-negative");

  const total = weights.reduce((sum, value) => sum + value, 0n);
  if (total <= 0n) throw new Error("weight sum must be positive");

  const numerators = weights.map((value) => value * scale);
  const floors = numerators.map((numerator) => numerator / total);
  let remaining = scale - floors.reduce((sum, value) => sum + value, 0n);
  const order = weights.map((_, index) => index).sort((left, right) => {
    const leftRemainder = numerators[left] % total;
    const rightRemainder = numerators[right] % total;
    if (leftRemainder === rightRemainder) return left - right;
    return leftRemainder > rightRemainder ? -1 : 1;
  });
  for (const index of order) {
    if (remaining === 0n) break;
    floors[index] += 1n;
    remaining -= 1n;
  }
  return floors;
}

export function matrixVectorProduct(
  matrix: readonly (readonly bigint[])[],
  vector: readonly bigint[],
): readonly bigint[] {
  if (matrix.length === 0 || vector.length === 0 || matrix.some((row) => row.length !== vector.length)) {
    throw new Error("matrix and vector dimensions must be non-empty and compatible");
  }
  return matrix.map((row) => row.reduce((sum, coefficient, index) => sum + coefficient * vector[index], 0n));
}

function normalizeJson(value: unknown): unknown {
  if (value === null || typeof value === "boolean" || typeof value === "string") return value;
  if (typeof value === "bigint") return value.toString(10);
  if (typeof value === "number") throw new Error("JavaScript numbers are forbidden in canonical AREKappa JSON");
  if (Array.isArray(value)) return value.map(normalizeJson);
  if (typeof value === "object") {
    const record = value as Record<string, unknown>;
    return Object.fromEntries(Object.keys(record).sort().map((key) => [key, normalizeJson(record[key])]));
  }
  throw new Error(`unsupported canonical JSON value: ${typeof value}`);
}

export function canonicalJson(value: unknown): string {
  return JSON.stringify(normalizeJson(value));
}
